src/components/cuda/linux-cuda.c

```diff
@@ -1,5651 +1,442 @@
 /**
  * @file    linux-cuda.c
- * @author  Tony Castaldo tonycastaldo@icl.utk.edu (updated in 08/2019, to make counters accumulate.)
- * @author  Tony Castaldo tonycastaldo@icl.utk.edu (updated in 2018, to use batch reads and support nvlink metrics.
- * @author  Asim YarKhan yarkhan@icl.utk.edu (updated in 2017 to support CUDA metrics)
- * @author  Asim YarKhan yarkhan@icl.utk.edu (updated in 2015 for multiple CUDA contexts/devices)
- * @author  Heike Jagode (First version, in collaboration with Robert Dietrich, TU Dresden) jagode@icl.utk.edu
+ * @author  Anustuv Pal
+ *          anustuv@icl.utk.edu
  *
  * @ingroup papi_components
  *
- * @brief This implements a PAPI component that enables PAPI-C to
- *  access hardware monitoring counters for NVIDIA CUDA GPU devices
- *  through the CUPTI library.
+ * @brief
+ *  This file implements a PAPI component that enables PAPI-C to access
+ *  hardware monitoring counters for NVIDIA GPU devices through the CuPTI library.
  *
  * The open source software license for PAPI conforms to the BSD
  * License template.
  */
 
-//-----------------------------------------------------------------------------
-// A basic assumption here (and in other components) is that we put as much of
-// the computational load of this component into the initialization stage and
-// the "adding" stage for events (update_control), because users are likely not
-// measuring performance at those times, but may well be reading these events
-// when performance matters. So we want the read operation lightweight, but we
-// can remember tables and such at startup and when servicing a PAPI_add().
-//-----------------------------------------------------------------------------
+#include <papi.h>
+#include <papi_internal.h>
+#include <papi_vector.h>
 
-#include <dlfcn.h>
-#include <limits.h>
-#include <float.h> // For DBL_MAX. 
-#include <sys/stat.h>
-#include <dirent.h>
+#include <string.h>
 
-// NOTE: We can't use extended directories; these include files have includes.
-#include <cupti.h>
-#include <cuda_runtime_api.h>
+#include "lcuda_common.h"
```

Comment:
Because it's being used.

Reply:
Why is it being used?

Reply:
`lcuda_common.h` contains API for dynamic event table. It is used by `linux-cuda.c` as well as `cupti_profiler.c`. `cupti_events.c` should also use this.

Reply:
Sharing the a global event table between frontend and backend is not a good idea. The only interaction between frontend and backend should be the main interface.

Reply:
Have a look at the API you will understand. This header is not itself responsible for passing the global table to the backend. It only contains the data struct and accompanying functions to support dynamic event table creation, deletion, insertion, and search. That facility is shared by both the layers.

In case of enumeration, the frontend has to pass the pointer to the global event table to the backend, right? As in
`cuptid_event_enum(global_event_table)`. How else will the global table be populated?

I can change it so that we are not passing `global_event_table` and `event_ids` to `cuptid_control_create`.
Is that something you prefer?

Reply:
I guess that is ok, you are right. My reasoning was that the backend and the frontend are completely independent. They share nothing and interact only through the cupti wrapper API. However, the backend already uses definitions from PAPI, like error codes. Could you just rename `event_list_t` to `ntv_event_table_t`? List in the name reveals implementation details for the data structure. `event_rec_t` can also be just `ntv_event_t`. So that the `ntv_event_table_t` is a collection of `ntv_event_t` elements.

Reply:
Another common field is `evt_pos`. It is used by the low end to transport measured values back to the correct position of the event that it is measuring for.